primitives/src/side_effect/mod.rs

```diff
@@ -86,22 +95,177 @@ impl Default for SecurityLvl {
     }
 }
 
-#[derive(Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
 pub struct FullSideEffect<AccountId, BlockNumber, BalanceOf> {
     pub input: SideEffect<AccountId, BlockNumber, BalanceOf>,
     pub confirmed: Option<ConfirmedSideEffect<AccountId, BlockNumber, BalanceOf>>,
     pub security_lvl: SecurityLvl,
 }
 
+impl<
+        AccountId: Encode,
+        BlockNumber: Ord + Copy + Zero + Encode,
+        BalanceOf: Copy + Zero + Encode + Decode,
+    > FullSideEffect<AccountId, BlockNumber, BalanceOf>
+{
+    pub fn harden(
+        &self,
+    ) -> Result<
+        (
+            SecurityLvl,
+            ConfirmationOutcome,
+            CircuitBalance,
+            AccountId32,
+            CircuitBlockNumber,
+            CircuitBalance,
+            [u8; 4],
+            Vec<Bytes>,
+            [u8; 4],
+        ),
+        &'static str,
```

Comment:
This should really use an Error enum defined in this module, same for the other errors

Reply:
will do